concordia/language_model/local_models.py

```diff
@@ -0,0 +1,307 @@
+        streaming: bool = False,
+    ):
+        """Initializes the instance.
+
+        Args:
+          model_name: The language model to use.
+          measurements: The measurements object to log usage statistics to.
+          channel: The channel to write the statistics to.
+        """
+        self._model_name = model_name
+        self._measurements = measurements
+        self._channel = channel
+        if streaming:
+            self._client = Ollama(
+                model=model_name, 
+                callback_manager=CallbackManager([StreamingStdOutCallbackHandler()])
+            )
+        else:
+            self._client = Ollama(model=model_name)
+
+    @override
+    def sample_text(
+        self,
+        prompt: str,
+        *,
+        max_tokens: int = language_model.DEFAULT_MAX_TOKENS,
+        max_characters: int = language_model.DEFAULT_MAX_CHARACTERS,
+        terminators: Collection[str] = language_model.DEFAULT_TERMINATORS,
+        temperature: float = language_model.DEFAULT_TEMPERATURE,
+        timeout: float = language_model.DEFAULT_TIMEOUT_SECONDS,
+        seed: int | None = None,
+        system_message = TEXT_SYSTEM_MESSAGE,
+    ) -> str:
+        
+        message = prompt
+        prompt = PROMPT_TEMPLATE.substitute(system_message=system_message, message=message)
+
+        self.logger.debug(f"Sending prompt to LLM: {prompt}")
+        for retry in range(_MAX_SAMPLE_TEXT_ATTEMPTS): 
+            try:
+                response = self._client(
+                    prompt,
+                    stop=terminators,
+                    temperature=temperature,
+                )
+            except ValueError as e:
+                self.logger.error(f"Error while calling LLM with input: {prompt}. Attempt {retry+1} failed.")
+                if retry == _MAX_SAMPLE_TEXT_ATTEMPTS - 1: 
+                    self.logger.error(f"Max retries exceeded. Raising exception.")
+                    raise language_model.InvalidResponseError(prompt)
+            else:
+                self.logger.debug(f"Succeeded after {retry+1} attempts.")
+                self.logger.debug(f"Response from LLM: {response}")
+                break
+
+        if self._measurements is not None:
+            self._measurements.publish_datum(
+                self._channel,
+                {'raw_text_length': len(response)},
+            )
+        return response
+
+    @staticmethod
+    def extract_choices(text):
+        match = re.search(r'\(?(\w)\)', text)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    @override
+    def sample_choice(
+        self,
+        prompt: str,
+        responses: Sequence[str],
+        *,
+        seed: int | None = None,
+    ) -> tuple[int, str, dict[str, float]]:
+        max_characters = len(max(responses, key=len))
+
+        attempts = 1
+        for _ in range(_MAX_MULTIPLE_CHOICE_ATTEMPTS):
+            sample = self.sample_text(
```

Comment:
This will sample _MAX_MULITPLE_CHOICE_ATTEMPTS * _MAX_SAMPLE_TEST_ATTEMPT = 25 times. Is that intentional? Is it necessary or can you just adjust some timeout?

FYI: The retry wrapper can avoid this?

Reply:
removed from this committ